### PR TITLE
Add Registry model package API

### PR DESCRIPTION
**Pull request by chaoyu-msft**
# Data Plane API - Pull Request

## API Info: The Basics
Most of the information about your service should be captured in the issue that serves as your [*API Spec engagement record*](https://aka.ms/azsdk/onboarding/restapischedule).

* Link to API Spec engagement record issue:

Is this review for (select one):

- [ ] a private preview
- [ ] a public preview
- [x] GA release

### Change Scope

This is a PR to merge an existing preview API into a GA dev branch. 
Specifically, it is to bring our Registry ModelVersion package API in 2024-04-01-preview version into 2024-04-01 GA branch.

<sup>This section will help us focus on the specific parts of your API that are new or have been modified. <br/>Please share a link to the design document for the new APIs, a link to the previous API Spec document (if applicable), and the root paths that have been updated. </sup>
* Design Document:
* Previous API Spec Doc:
* Updated paths:

### Viewing API changes

For convenient view of the API changes made by this PR, refer to the URLs provided in the table 
in the `Generated ApiView` comment added to this PR. You can use ApiView to show API versions diff. 

### Suppressing failures

If one or multiple validation error/warning suppression(s) is detected in your PR, please follow the 
[Swagger-Suppression-Process](https://aka.ms/pr-suppressions) 
to get approval.

## ❔Got questions? Need additional info?? We are here to help!

<details>
  <summary> Contact us!</summary>

The [Azure API Review Board](https://aka.ms/azsdk/onboarding/restapischedule) is dedicated to helping you create amazing APIs. You can read about our mission and learn more about our process on our [wiki](https://aka.ms/azsdk/onboarding/restapischedule).
* 💬 [Teams Channel](https://teams.microsoft.com/l/channel/19%3a3ebb18fded0e47938f998e196a52952f%40thread.tacv2/General?groupId=1a10b50c-e870-4fe0-8483-bf5542a8d2d8&tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47)
* 💌 [email](mailto://azureapirbcore@microsoft.com)

</details>

<details>
  <summary>Click here for links to tools, specs, guidelines & other good stuff</summary>

### Tooling

 * [Open API validation tools](https://aka.ms/swaggertools) were run on this PR. Go here to see [how to fix errors](https://aka.ms/ci-fix)
 * [Spectral Linting](https://github.com/Azure/azure-api-style-guide/blob/main/README.md)
 * [Open API Hub](https://aka.ms/openapiportal)

### Guidelines & Specifications

 * [Azure REST API Guidelines](https://aka.ms/azapi/guidelines)
 * [OpenAPI Style Guidelines](https://aka.ms/azapi/style)
 * [Azure Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy)

### Helpful Links

 * [Azure DevTools Wiki](https://aka.ms/azapi)

</details>

<details>
  <summary>Checks stuck in `queued` state?</summary>
If the PR CI checks appear to be stuck in `queued` state, please add a comment with contents `/azp run`.
This should result in a new comment denoting a `PR validation pipeline` has started and the checks should be updated after few minutes.  
</details>

### CODE REVIEW

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json [hunk 1]
```diff
@@ -3261,6 +3261,98 @@
...
+        "operationId": "RegistryModelVersions_Package",
+        "consumes": [
+          "application/json"
+        ],
+        "produces": [
+          "application/json"
+        ],
+        "parameters": [
+          {
+            "$ref": "../../../../../common-types/resource-management/v3/types.json#/parameters/SubscriptionIdParameter"
```
Comment (mentat9) [inline specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json:3279]: >v3 [](http://example.com/codeflow?start=69&length=2)

Should use v5. Comment applies throughout.

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json [hunk 2]
```diff
@@ -3261,6 +3261,98 @@
         }
       }
     },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.MachineLearningServices/registries/{registryName}/models/{modelName}/versions/{version}/package": {
```
Comment (mentat9) [inline specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json:3264]: >package [](http://example.com/codeflow?start=180&length=7)

Although package is a verb, it is very commonly a noun. Consider using a more action-y sounding name like buildPackage, makePackage, createPackate, or similar to more strongly indicate performing an action on a resource.

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json [hunk 3]
```diff
@@ -19466,11 +19813,374 @@
...
+          "x-ms-mutability": [
+            "create",
+            "read"
+          ],
+          "x-nullable": true
+        },
+        "environmentVariables": {
+          "description": "Collection of environment variables.",
+          "type": "object",
+          "additionalProperties": {
```
Comment (mentat9) [inline specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json:19979]: >additionalProperties [](http://example.com/codeflow?start=11&length=20)

This is an ARM anti-pattern and generally not allowed. Can you model this as an array of (string, string) tuples?

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json [hunk 4]
```diff
@@ -19466,11 +19813,374 @@
...
+          "x-ms-mutability": [
+            "create",
+            "read"
+          ],
+          "x-nullable": true
+        },
+        "properties": {
+          "description": "Property dictionary. Properties can be added, removed, and updated.",
+          "type": "object",
+          "additionalProperties": {
```
Comment (mentat9) [inline specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json:20022]: >additionalProperties [](http://example.com/codeflow?start=11&length=20)

ARM anti-pattern not allowed. Please schematize this object or model it as an array.

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json [hunk 5]
```diff
@@ -19466,11 +19813,374 @@
...
+            "x-nullable": true
+          },
+          "x-ms-mutability": [
+            "create",
+            "read"
+          ],
+          "x-nullable": true
+        },
+        "targetEnvironmentId": {
+          "description": "[Required] Arm ID of the target environment to be created by package operation.",
```
Comment (mentat9) [inline specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json:20046]: >Arm [](http://example.com/codeflow?start=37&length=3)

Don't use ARM in customer-visible names or text. Prefer Azure instead.

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json [hunk 6]
```diff
@@ -19466,11 +19813,374 @@
...
+          "minLength": 1,
+          "pattern": "[a-zA-Z0-9_]",
+          "type": "string",
```
Comment (mentat9) [inline specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json:20049]: >type": "string [](http://example.com/codeflow?start=11&length=14)

format: arm-id (+ optional collection of valid resource types)?

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json [hunk 7]
```diff
@@ -19334,6 +19645,42 @@
...
       "additionalProperties": false
     },
+    "OnlineInferenceConfiguration": {
+      "description": "Online inference configuration options.",
+      "type": "object",
+      "properties": {
+        "configurations": {
+          "description": "Additional configurations",
+          "type": "object",
+          "additionalProperties": {
```
Comment (mentat9) [inline specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json:19655]: >additionalProperties [](http://example.com/codeflow?start=11&length=20)

ARM anti-pattern not allowed. Please schematize this object or model it as an array.

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json [hunk 8]
```diff
@@ -19466,11 +19813,374 @@
...
+      "type": "object",
+      "allOf": [
+        {
+          "$ref": "#/definitions/PackageInputPathBase"
+        }
+      ],
+      "properties": {
+        "url": {
+          "description": "Input path url.",
+          "type": "string",
```
Comment (mentat9) [inline specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json:19908]: >"type": "string" [](http://example.com/codeflow?start=10&length=16)

Could use format: uri for this.

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json [hunk 9]
```diff
@@ -11882,6 +12010,58 @@
...
+        {
+          "$ref": "#/definitions/BaseEnvironmentSource"
+        }
+      ],
+      "properties": {
+        "resourceId": {
+          "description": "[Required] Resource id accepting ArmId or AzureMlId.",
+          "minLength": 1,
+          "pattern": "[a-zA-Z0-9_]",
+          "type": "string"
```
Comment (mentat9) [inline specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json:12029]: >"type": "string" [](http://example.com/codeflow?start=10&length=16)

Consider using format arm-id: and [optionally] a set of allowed resource types. Comment applies throughout.

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json [hunk 10]
```diff
@@ -19466,11 +19813,374 @@
...
+      "type": "object",
+      "allOf": [
+        {
+          "$ref": "#/definitions/PackageInputPathBase"
+        }
+      ],
+      "properties": {
+        "resourceId": {
+          "description": "Input resource id.",
+          "type": "string",
```
Comment (mentat9) [inline specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json:19890]: >"type": "string" [](http://example.com/codeflow?start=10&length=16)

format: arm-id?

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json [hunk 11]
```diff
@@ -19466,11 +19813,374 @@
...
+          "$ref": "#/definitions/PackageBuildState",
+          "readOnly": true,
+          "x-ms-mutability": [
+            "read"
+          ]
+        },
+        "environmentVariables": {
+          "description": "Collection of environment variables.",
+          "type": "object",
+          "additionalProperties": {
```
Comment (mentat9) [inline specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json:20091]: >additionalProperties [](http://example.com/codeflow?start=11&length=20)

ARM anti-pattern not allowed. Please schematize this object or model it as an array.

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json [hunk 12]
```diff
@@ -19466,11 +19813,374 @@
...
+          "readOnly": true,
+          "x-ms-mutability": [
+            "read"
+          ],
+          "x-nullable": true,
+          "x-ms-identifiers": []
+        },
+        "logUrl": {
+          "description": "Log url of the image build operation.",
+          "type": "string",
```
Comment (mentat9) [inline specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json:20125]: >"type": "string" [](http://example.com/codeflow?start=10&length=16)

format: uri?

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json [hunk 13]
```diff
@@ -19466,11 +19813,374 @@
...
+          "readOnly": true,
+          "x-ms-mutability": [
+            "read"
+          ],
+          "x-nullable": true
+        },
+        "properties": {
+          "description": "Property dictionary. Tags can be added, removed, and updated.",
+          "type": "object",
+          "additionalProperties": {
```
Comment (mentat9) [inline specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json:20144]: >additionalProperties [](http://example.com/codeflow?start=11&length=20)

ARM anti-pattern not allowed. Please schematize this object or model it as an array.

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json (hunk 9 above)
Comment (chaoyu-msft) [reply specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json:12029]: This value accepts both arm-id and AzureMLId.

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json [hunk 14]
```diff
@@ -19466,11 +19813,374 @@
...
+            "type": "string",
+            "x-nullable": true
+          },
+          "x-ms-mutability": [
+            "create",
+            "read"
+          ],
+          "x-nullable": true
+        },
+        "tags": {
```
Comment (mentat9) [inline specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json:20032]: >tags [](http://example.com/codeflow?start=9&length=4)

What is the scenario for implementing your own tags? Is there a good reason, this couldn't be implemented as a tracked resource, so the tags feature is aligned with ARM?

**Comment by chaoyu-msft** [comment]
LintDiff check is currently broken, author is required to execute it locally and paste the result for review.
Reference: https://github.com/Azure/azure-openapi-validator/blob/main/CONTRIBUTING.md#how-to-run-lintdiff-locally-from-source

**Comment by chaoyu-msft** [comment]
Closing this PR, the API author will continue working on this change.

**Pull request closed without merging**